### PR TITLE
Support flattened fields and multi-fields as dimensions in downsampling

### LINKED ISSUE #99297: (Possible) missing downsampling test with flattened field dimension
### Elasticsearch Version

8.7.0 and above

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

All

### Problem Description

Make sure we have a test running downsampling on a TSDB index that defines dimensions including at least a flattened field.
Note that using a flattened field as dimension requires dimension fields to be explicitly defined by means of an array `time_series_dimensions: [dim_1, dim_2,...]` where each dimension path is taken from the root of the flattened field.

### Steps to Reproduce

Try to downsample a TSDB index that uses a flattened field as dimension.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by lkts**
Closes #99297.

### CODE REVIEW

File: x-pack/plugin/downsample/src/test/java/org/elasticsearch/xpack/downsample/DownsampleActionSingleNodeTests.java [hunk 1]
```diff
@@ -212,6 +214,12 @@ public void setup() throws IOException {
         // Dimensions
         mapping.startObject(FIELD_DIMENSION_1).field("type", "keyword").field("time_series_dimension", true).endObject();
         mapping.startObject(FIELD_DIMENSION_2).field("type", "long").field("time_series_dimension", true).endObject();
+        mapping.startObject(FIELD_DIMENSION_3).field("type", "flattened").array("time_series_dimensions", "level1_value", "level1_obj.level2_value").endObject();
```
Comment (lkts) [inline x-pack/plugin/downsample/src/test/java/org/elasticsearch/xpack/downsample/DownsampleActionSingleNodeTests.java:217]: Couldn't find lower level unit tests so added them here.

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DimensionFieldValueFetcher.java [hunk 2]
```diff
@@ -42,12 +42,18 @@ static List<FieldValueFetcher> create(final SearchExecutionContext context, fina
...
-            if (context.fieldExistsInIndex(dimension)) {
+            if (context.fieldExistsInIndex(fieldType.name())) {
                 final IndexFieldData<?> fieldData = context.getForField(fieldType, MappedFieldType.FielddataOperation.SEARCH);
-                final String fieldName = context.isMultiField(dimension)
-                    ? fieldType.name().substring(0, fieldType.name().lastIndexOf('.'))
-                    : fieldType.name();
-                fetchers.add(new DimensionFieldValueFetcher(fieldType, fieldData));
+                if (fieldType instanceof FlattenedFieldMapper.KeyedFlattenedFieldType flattenedFieldType) {
+                    // Name of the field type and name of the dimension are different in this case.
+                    var dimensionName = flattenedFieldType.rootName() + '.' + flattenedFieldType.key();
```
Comment (lkts) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DimensionFieldValueFetcher.java:49]: Can you come up with a scenario when flattened field can be a multi-field? Cause right now this does not support that.

**Comment by lkts** [comment]
@elasticmachine update branch

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DimensionFieldValueFetcher.java (hunk 2 above)
Comment (martijnvg) [reply x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DimensionFieldValueFetcher.java:49]: I can't think of a scenario. A flattened field isn't a leaf field, but an object field that gets indexed as if it is a leaf field. So you can indexing it into a single numeric and keyword field via multi fields isn't possible.

**Review by martijnvg — APPROVED** [review]
Thanks for working on this!

**Pull request merged**